### ISSUE TITLE
GDAL_Translate applies target formatting parameters to source extent before evaluating projwin

### ISSUE THREAD

**Issue by BR-dev0** [OSGeo/gdal#6284]
The following behavior was experienced with GDAL 3.5.1 on Win10x64 (GIS-Internals build VS2019) and tested with older GDAL 3.4.0 to be no new issue:

I've got a series of 5000+ *.jpg aerial tiles, which need to be translated to different tile extents, target resolution and compression quality. A source.vrt was created from the aerials with gdalbuildvrt. A tiles.csv contains rows with names and extent parameters for the target tiles.

A (cmd) FOR-Loop through the rows of tiles.csv requires gdal_translate around 4 hours(!) for the creation of _each_ new tile, when reformatting information is included:
```
for /f "skip=1 delims=; tokens=1-5" %%i in (tiles.csv) do (
  gdal_translate -projwin %%j %%m %%l %%k -tr 0.4 0.4 -r cubicspline -scale -co QUALITY=75 source.vrt %%i.jpg
  )
```

When omitting the reformatting parameters, _each_ new tile is completed after around 2 minutes:
```
for /f "skip=1 delims=; tokens=1-5" %%i in (tiles.csv) do (
  gdal_translate -projwin %%j %%m %%l %%k source.vrt %%i.jpg
  )
```

When source.vrt is limited to 20 of the aerial images, _all_ of them are translated (_including_ the reformatting parameters) in about 1 minute. 

Therefore, evaluating the source.vrt for all 5000+ aerial images and identifying the necessary images for projwin takes gdal_translate nearly 2 minutes. Which seems acceptable, considering source.vrt is about 40 MByte in size. The tiling and writing process as well as the reformatting is done in mere seconds for each tile, when the source.vrt extent is limited.

The only explanation for the long duration of tiling _and_ reformatting of the source.vrt for all 5000+ images seems to be, that reformatting is applied to each source aerial included in the source.vrt, _before_ the projwin parameters are evaluated. Monitoring the disk access (which is continuosly high during the 4 hours per tile on the source drive, but neither the target drive (containing the source.vrt and - eventually - the target tiles) nor the Windows system drive (containing the GDAL installation as well as the %temp% folder), seems to confirm this assumption. (All 3 drives are physically different.)

I'd expect gdal_translate to do this the other way round: _First_ creating a new tile from the source tiles and _afterwards_ reformatting it to disk.

**Comment by rouault** [comment]
> I'd expect gdal_translate to do this the other way round: _First_ creating a new tile from the source tiles and _afterwards_ reformatting it to disk.

Explaining how gdal_translate and the underlying logic works is complicated, but basically everything is done at the same time to avoid creating temporary files.

Anyway your ticket is not really about a bug but how to optimize GDAL usage and it is not the best forum for that (we have already enough clearly "unitary" reproducable issues to deal with, to not be able to tackle complex use cases like yours).
potential suggestion: use ``gdal_translate each_source_jpeg.jpg temporary_tiff.tif -co COMPRESS=JPEG -co TILED=YES`` to create tiled compress GeoTIFF (this will convert in a lossless way your original JPEG) , and use them to build the VRT.

**Comment by BR-dev0** [comment]
Thank you very much for your quick reply!

The *.vrt is only an intermediate step to have a unified source prior to re-tiling. The individual target *.jpg tiles will be divided between users for their work areas. The programs they use for visualizing this data don't support *.vrt (mostly ESRI ArcPad on old GNSS Windows Mobile devices with very low memory and computing power) and include scripts to auto-load the tiles. The tiles must match the names and extents of other tiles meant for overlaying. Therefore, GeoTIFF and *.vrt from GeoTIFF, unfortunately, get me nowhere.

> Explaining how gdal_translate and the underlying logic works is complicated, but basically everything is done at the same time to avoid creating temporary files. [...] Anyway your ticket is not really about a bug but how to optimize GDAL usage

Hm. The amount of (as of the target extent) unnecessary background calculation is so massive, that I'd call the evaluation algorithm buggy. But it is your code and deep insight. So I, of course, accept your assessment. (And *.vrt probably is not the main input considered for gdal_translate, anyways...)

If gdal_translate will not be optimized for first evaluating the target extent (in near future), I'll probably use the unreformatted tiling method (I mentioned above) for first creating the target images in the target extents. Afterwards, I'll reformat them in a separate step with either gdal_translate or imagemagick. _Sigh._

As a last ressort, I can go back to ESRI ArcGIS SplitRaster_management() with split_method=POLYGON_FEATURES. This works fine and sufficiently fast and can be scripted from command line on machines with a registered copy of ArcGIS. I just hoped to be able to finally replace this with Open Source command line tools. (I'm aware of Murat Çalışkans Easy Raster Splitter QGIS plugin.)

**Comment by rouault** [comment]
There might be a inefficiency somewhere. Would be easier if you could provide a *minimum* standalone reproducer (as small as possible, but still large enough so that it can be timed accurately enough, e.g something that takes at least ~ 500 ms in the slow case) so that someone can at least try to reproduce easily

**Comment by BR-dev0** [comment]
With too few images, the effect is overshadowed by gdal_translate initialization time. And the delay increases overproportional with the number of images included in the *.vrt.

Here are the results when executed with 30 and 60 aerials included in the *.vrt, which acts as source file for gdal_translate:
```
30 aerials -> start: 15:53:09,79 | end: 15:53:17,98 (= 8,19 seconds)
60 aerials -> start: 15:53:17,99 | end: 15:53:29,91 (= 11,92 seconds)
```

Although the numbers are dominated by gdal_translate initialization, it is clear, that it shouldn't take more than 3 seconds to identify the (same) 9 aerials required to create the target tile within the double-sized (but still very small) *.vrt.

Source of the aerials is https://data.geobasis-bb.de/geobasis/daten/dop/rgb_jpg. You can create batch files to download certain areas on this site or individually load single images (as *.zip). For above test I used about half of one of the smallest areas (capital region Potsdam): 

[get_DOP.txt](https://github.com/OSGeo/gdal/files/9462272/get_DOP.txt)

All aerials are extracted to .\DOP_Source and half of the aerials (those ending on ..11 to ..20) are extracted to .\DOP_Source_half.
Afterwards, a *.vrt is created to .\DOP_Target\Source.vrt and .\DOP_Target_half\Source_half.vrt, respectively. The following go.cmd file extracted 1 (identical) tile from each source with the above mentioned timings:

```
@echo off & setlocal enabledelayedexpansion enableextensions
echo !time! && gdal_translate --config CPL_DEBUG ON -projwin 358000 5817000 360500 5814500 -tr 0.4 0.4 -r cubicspline -scale -co QUALITY=75 .\DOP_Target_half\Source_half.vrt .\DOP_Target_half\DOP_test_half.jpg && echo !time!
echo:
echo !time! && gdal_translate --config CPL_DEBUG ON -projwin 358000 5817000 360500 5814500 -tr 0.4 0.4 -r cubicspline -scale -co QUALITY=75 .\DOP_Target\Source.vrt .\DOP_Target\DOP_test.jpg && echo !time!
```

Here is the log file:

[gdal_translate.log](https://github.com/OSGeo/gdal/files/9462369/gdal_translate.log)

If you get all aerials from above mentioned site, extracting the example tile would probably take a couple of hours for you, as well. – Or you have a _real_ powerful machine with several hundred GByte RAM...  ;-)

**Comment by jratike80** [comment]
I suppose there is now something that is specific to using JPEG images as sources. We use VRT files which combines thousands of GeoTIFF images with good success in a process that extracts pieces of DEM for our users. VRT is probably not the most effective format because it does not support spatial indexing but the overhead is just in parsing the  whole XML file each time. At least I believe that is what happens.

**Comment by BR-dev0** [comment]
> I suppose there is now something that is specific to using JPEG images as sources. We use VRT files which combines thousands of GeoTIFF images with good success in a process that extracts pieces of DEM for our users. VRT is probably not the most effective format because it does not support spatial indexing but the overhead is just in parsing the whole XML file each time. At least I believe that is what happens.

No. The Problem is the difference in handling between just extracting a target tile defined by projwin and also reformatting the output. When I change the above go.cmd this way, omitting the reformatting parameters:

```
@echo off & setlocal enabledelayedexpansion enableextensions
echo !time! && gdal_translate --config CPL_DEBUG ON -projwin 358000 5817000 360500 5814500 .\DOP_Target_half\Source_half.vrt .\DOP_Target_half\DOP_test_half_noconvert.jpg && echo !time!
echo:
echo !time! && gdal_translate --config CPL_DEBUG ON -projwin 358000 5817000 360500 5814500 .\DOP_Target\Source.vrt .\DOP_Target\DOP_test_noconvert.jpg && echo !time!
```

then the timing changes to:

```
30 aerials -> start: 07:50:19,65 | end: 07:50:27,91 (= 8,26 seconds)
60 aerials -> start: 07:50:27,92 | end: 07:50:36,12 (= 8,20 seconds)
```

This means, without reformatting, there's virtually no difference in handling *.vrt files linking to 30 or 60 aerial images. The following log file reveals the reason, which I already pointed out in the subject of this thread and my first posts:

[gdal_translate_noconvert.log](https://github.com/OSGeo/gdal/files/9475206/gdal_translate_noconvert.log)

Without reformatting, only the 9 source tiles necessary for writing the target tile (described by projwin) are accessed on disk. With reformatting, OTOH, each of the 30 or 60 source aerials on disk is not only accessd, but (judging from the timing information I provided in post 5 above) also (completely unnecessarily!) reformatted,

**Comment by jratike80** [comment]
Ok, understood. You indeed explained that in your report, sorry for not reading it thoroughly.
If you could somehow create a temporary VRT that contains just the source images that are needed then the ` -tr 0.4 0.4 -r cubicspline -scale -co QUALITY=75` would be fast.
What happens with
 `gdal_translate --config CPL_DEBUG ON -projwin 358000 5817000 360500 5814500 .\DOP_Target_half\Source_half.vrt .\DOP_Target_half\DOP_test_half_noconvert.vrt`
 Does it copy the whole huge VRT and limit the extent only with SrcRect? Does applying the advanced paraters to this new VRT also trigger the scanning of all the 5000 originals?

**Comment by BR-dev0** [comment]
> If you could somehow create a temporary VRT that contains just the source images that are needed then the `-tr 0.4 0.4 -r cubicspline -scale -co QUALITY=75` would be fast.
> What happens with
> `gdal_translate --config CPL_DEBUG ON -projwin 358000 5817000 360500 5814500 .\DOP_Target_half\Source_half.vrt .\DOP_Target_half\DOP_test_half_noconvert.vrt`
> Does it copy the whole huge VRT and limit the extent only with SrcRect? Does applying the advanced paraters to this new VRT also trigger the scanning of all the 5000 originals?

This approach references from the subselect.vrt to a sub-area in the original source.vrt. When using subselect.vrt as source image for gdal_translate _with_ fromatting parameters, only the 9 necessary *.jpg source aerial images are accessed on disk, though. The go.cmd changes to:

```
@echo off & setlocal enabledelayedexpansion enableextensions
echo !time! && gdal_translate --config CPL_DEBUG ON -projwin 358000 5817000 360500 5814500 .\DOP_Target_half\Source_half.vrt .\DOP_Target_half\DOP_Subselect_half.vrt && echo !time!
echo:
echo !time! && gdal_translate --config CPL_DEBUG ON -projwin 358000 5817000 360500 5814500 .\DOP_Target\Source.vrt .\DOP_Target\DOP_Subselect.vrt && echo !time!
echo:
echo !time! && gdal_translate --config CPL_DEBUG ON -tr 0.4 0.4 -r cubicspline -scale -co QUALITY=75 .\DOP_Target_half\DOP_Subselect_half.vrt .\DOP_Target_half\DOP_test_half.jpg && echo !time!
echo:
echo !time! && gdal_translate --config CPL_DEBUG ON -tr 0.4 0.4 -r cubicspline -scale -co QUALITY=75 .\DOP_Target\DOP_Subselect.vrt .\DOP_Target\DOP_test.jpg && echo !time!
```

Timing ([gdal_translate_subselect.log](https://github.com/OSGeo/gdal/files/9476865/gdal_translate_subselect.log)) looks promising:

```
Preparation (create subset.vrt):
30 aerials -> start: 11:26:38,03 | end: 11:26:38,65 (= 0,62 seconds)
60 aerials -> start: 11:26:38,65 | end: 11:26:39,28 (= 0,63 seconds)
Execution (apply subset.vrt as source):
30 aerials -> start: 11:26:39,29 | end: 11:26:44,48 (= 5,19 seconds)
60 aerials -> start: 11:26:44,49 | end: 11:26:49,67 (= 5,18 seconds)
```

Although being a two-step process, this extraction is even faster than extracting directly _without_ reformatting! I have yet to see, how this scales to the whole set of aerials. But it sure looks better than first extracting all target images without reformatting and reformatting later in a second step. (Which I currently considered as a possible workaround.) – Thanks a lot for this suggestion!!

When I decided to look for an Open Source replacement for ESRI ArcGIS SplitRaster_management() with split_method=POLYGON_FEATURES I read through several decades of information on the Internet. Since at least 25 years people state, that they miss a possibility to retile a set of images to different extents by a polygon file containing not only the extents of the target tiles, but also the target file names from an attribute column. The closest Open Source solution I came across was the (relatively new) Easy Raster Splitter QGIS plugin by Murat Çalışkans. But that's a GUI solution and therefore not to be easily included in a (much larger) comand line data preparation workflow.

The next best suggestion was scripting gdal_translate in a FOR loop against a *.vrt of all source images and with projwin to define each target extent. This, unfortuately, has overhead (gdal_translate startup time and source.vrt read time) for each single iteration of the loop. But it generally _does_ work. It just has the additional problem when combined with reformatting, as decribed. If either of the two two-step methods works sufficiently fast enough (with your suggestion looking more promising), then I at least have a crutch replacement for the ESRI method...

**Comment by rouault** [comment]
I've managed to reproduce the issue with the data of https://github.com/OSGeo/gdal/issues/6284#issuecomment-1233022756

The reason for the performance issue is the use of the ``-scale`` switch. When used that way, it has to figure out the minimum and maximum value in the input dataset, that is inspect all pixels of a VRT file made of many files, and not just those intersecting -projwin. This behaviour makes sense (use min/max on the whole source file, not just the window of interest), otherwise you'd get tiles that wouldn't have consistent colorimetry
There's a simple solution to that performance issue: compute statistics with "gdalinfo -stats your.vrt", and they will be serialized in the VRT file, and further attempts to get min & max will be instant.
To be noted that in your use case, the source JPEG files uses the whole [0,255] dynamic, so the -scale parameter is just a no-operation, and you could just remove it

**Comment by BR-dev0** [comment]
Unexpected reason. But the explanation makes sense. Thank you very much!

**Comment by rouault** [comment]
documentation enhancement to capture this knowledge: https://github.com/OSGeo/gdal/pull/6305

**Comment by BR-dev0** [comment]
Additional thought: Couldn't the "scale" analysis stop, once the _maximum possible_ scale is encountered in the source file(s)?

**Comment by rouault** [comment]
> Additional thought: Couldn't the "scale" analysis stop, once the _maximum possible_ scale is encountered in the source file(s)?

that has been implemented in GDAL master a few weeks ago (when min=0 and max=255 for Byte data to be more exact). That helps a bit, but generally you use -scale for non-Byte data (typically 12-bit or 16-bit data where you rarely reach the minimum minimum and maximum maximum)

**Comment by BR-dev0** [comment]
Okay. Thank you for this information, again.

**Comment by BR-dev0** [comment]
For the record: Usefulness of the -scale parameter for my source data aside, I decided to give `gdalinfo -stats source.vrt` a try. It is running since more than 2 days (> 50 hours) in a row by now and shows no sign of potential completing. With SysInternals ProcMon I had a look, which aerials gdalinfo is accessing. It reads one aerial after another in 4096 byte chunks to completion and then continues to the next aerial. Depending on compression, it (completely) reads between 6 and 8 aerials per second and should have read all (nearly) 30.000 source aerials (5.000+ aerials is the target image number; I mixed these numbers up in my earlier posts) in about an hour. Extrapolation of these numbers means, the stats computation should have been completed since starting the command about 50 times. I'll leave the process running for another couple of hours, but eventually have to cancel computation sometime.

Seems like pre-running gdalinfo with stats computation isn't an applicable way to use -scale with huge *.vrt source files.

**Comment by rouault** [comment]
> gdalinfo with stats computation isn't an applicable way to use -scale with huge *.vrt source files.

https://github.com/OSGeo/gdal/pull/6349 should help

**Comment by BR-dev0** [comment]
The speed-optimized version required almost exactly 22 hours to complete `gdalinfo -stats source.vrt` for above mentioned source *.vrt referencing about 30.000 aerials. 

That's definitely a considerable speedup. How much so, I can't tell, because I killed the process with the non-optimized gdalinfo program after 77 hours (with no sign of nearing completion). I had also tried running with -approx_stats parameter, but killed that process after 60 hours.

Thank you very much for looking into this issue!
